core/src/main/scala/kafka/log/Log.scala

```diff
@@ -497,8 +574,12 @@ class Log(@volatile var dir: File,
     var monotonic = true
     var maxTimestamp = RecordBatch.NO_TIMESTAMP
     var offsetOfMaxTimestamp = -1L
+    var isDuplicate = false
+    val producerAppendInfos = mutable.Map[Long, ProducerAppendInfo]()
 
     for (batch <- records.batches.asScala) {
+      if (isFromClient && batch.magic >= RecordBatch.MAGIC_VALUE_V2 && shallowMessageCount > 0)
+        throw new InvalidRecordException("Client produce requests should not have more than one batch")
```

Comment:
Hmm, ProducerRequest.validateRecords() is called on the broker side when converting bytes from socket to a request object. So, even if librd client has an issue, the broker should still be able to capture this when constructing ProducerRequest.